Attempt to recover from isolated orphaning

hashmap opened the pull request:
Due to occasional bad network conditions, a node can miss a block.
When the next one is received and detected orphaned, we should at least try a request for the
previous block, assuming it's not already an orphan in itself and we're not syncing already.

No additinal tests were implemented, tests like `simulate_full_sync` test this functionality because
sync starts later. It seems to be an issue with sync test coverage. As fix I'd suggest to start sync
as soon as we get a peer connection.

[#705]

ignopeverell commented on grin/src/adapters.rs:
```diff
@@ -376,20 +375,33 @@ impl NetToChainAdapter {
+                Ok(_) => true,
+                Err(e) => {
+			        debug!(LOGGER, "Block {} refused by chain: {:?}", bhash, e);
+                    match e {
```
I would merge that match into the one above, the debug statement can just be moved into the `true` case as the other branches are already covered.

hashmap replied:
Tried to preserve the original log message, but you are right, we have specific ones in each case

ignopeverell commented on grin/src/adapters.rs:
```diff
@@ -88,9 +88,8 @@ impl p2p::ChainAdapter for NetToChainAdapter {
-
-		self.process_block(b)
-	}
+        self.process_block(b, addr)
```
Looks like your indentations are off.

hashmap replied:
Not just here, I'm a bit lost, it looks like 8-spaces tabs are used (I used 4), also do we replace tabs with spaces? Should be fixed in this PR anyway

ignopeverell commented on grin/src/adapters.rs:
```diff
@@ -398,16 +410,21 @@ impl NetToChainAdapter {
+	}
+
+	fn request_block_by_hash(&self, h: Hash, addr: &SocketAddr) {
+		if let None = wo(&self.peers).adapter.get_block(h.clone()) {
```
This loads a whole block, including parsing it, just to check if we already know about it. How about using `block_exists` instead of `get_block`?

hashmap commented:
I refactored block request part, originally thought it should be a separate pr, but since it was mentioned during review I implemented it. I found that existing code was ignoring too many error cases, so new version is much bigger

hashmap commented:
Not sure why it failed, will trigger another travis build.

ignopeverell commented:
Travis can be unreliable... Thanks for the great work!
